### ISSUE TITLE
Connect parameters not considered in cycleReferences cache

### ISSUE THREAD

**Issue by carlesarnal** [Apicurio/apicurio-registry#3606]
In `AvroData` with a struct like the following

```
{
        "type" : "struct",
        "fields" : [ {
          "type" : "int32",
          "optional" : false,
          "field" : "scale"
        }, {
          "type" : "bytes",
          "optional" : false,
          "field" : "value"
        } ],
        "optional" : true,
        "name" : "io.debezium.data.VariableScaleDecimal",
        "version" : 1,
        "doc" : "Variable scaled decimal",
        "parameters" : {
          "precision" : "10"
        },
        "field" : "VAL_F_10"
 }
```

There will be a cache hit in `fromConnectContext.cycleReferences` even if the parameters value is different. We must investigate using a different key (maybe the entire schema?).

**Comment by rk3rn3r** [comment]
Hey @carlesarnal, may I ask you what's the status of this one? We are getting some more complaints about this issue from Debezium users and while I am not working on a temporary workaround it would be good if that could be fixed soon. I might be able to look into it myself after I finished the workaround, but maybe you have a pointer for me? I think I look for `fromConnectContext.cycleReferences()` for the cache acccess?

**Comment by carlesarnal** [comment]
Hey,

No status, sorry. This is acknowledged but we haven't had the time to go through it yet. What you mention is the perfect start, unfortunately, I don't have any other insight into this one.

**Comment by mfvitale** [comment]
@carlesarnal any update on that?

**Comment by carlesarnal** [comment]
## Reopening — this was never actually fixed

The fix in PR #5637 (commit `bebcf82`) modified `fromConnectSchemaCache` to use `AvroDataSchemaCacheKey`, but that was the **wrong cache**. `ConnectSchema.equals()` already includes `parameters` and `defaultValue`, so `fromConnectSchemaCache` correctly distinguished schemas with different parameters. The fix was correctly reverted in #6248.

The actual bug — `FromConnectContext.cycleReferences` keyed by `schema.name()` only — was never addressed.

### Root cause

In `AvroData.java`:

- **Line 947**: `fromConnectContext.cycleReferences.put(schema.name(), baseSchema)` — stores by name only
- **Line 1099**: `fromConnectContext.cycleReferences.containsKey(schema.name())` — looks up by name only
- **Line 2470**: `Map<String, org.apache.avro.Schema> cycleReferences` — `String` key = schema name

When a single record has two fields with the same schema name but different `parameters` (e.g. two `io.debezium.data.VariableScaleDecimal` fields with different `precision` values), the second field hits the cycle cache and gets the first field's Avro schema with the wrong parameters.

### Fix needed

The `cycleReferences` key must incorporate `parameters` (and potentially `defaultValue`), not just the schema name. Options:

1. Key by `schema.name() + schema.parameters()` (a composite key)
2. Key by the full Connect `Schema` object (uses `ConnectSchema.equals()` which already includes parameters)

### Related

See also #7788 — a separate caching bug in the schema resolver's GAV index discovered during this investigation.

**Comment by carlesarnal** [comment]
**Correction to my earlier analysis**: the `AvroDataSchemaCacheKey` fix in PR #5637 likely **did** work around this bug in practice, even though it targeted the wrong cache (`fromConnectSchemaCache` instead of `cycleReferences`).

Here's why: by making `fromConnectSchemaCache` use a key that includes `parameters`, schemas with different parameters would miss the top-level cache and go through full `fromConnectSchema()` conversion independently. This meant `cycleReferences` (which is scoped to a single `FromConnectContext`) would be populated fresh for each conversion, avoiding the name-only collision.

When the fix was reverted in #6248, `fromConnectSchemaCache` went back to using `Schema` as the key. But `ConnectSchema.equals()` already includes `parameters`, so the top-level cache should still distinguish them... **unless** the issue is specifically about the `cycleReferences` within a single conversion call (e.g. Debezium envelope with `before`/`after` fields sharing the same schema name but different parameters within the same record).

The core bug remains: `cycleReferences` is `Map<String, Schema>` keyed by `schema.name()` only (line 2470). This needs to be fixed directly rather than worked around via the outer cache.

This is the bug currently being reported by Debezium users migrating to 3.x (see their report referencing the same `io.debezium.data.Enum` + `connect.parameters` pattern).
